lib/assert.js

```diff
@@ -931,6 +933,86 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class CallTracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      // If function has been called more than expected times, add back into
+      // callchecks.
+      if (context.actual === context.exact + 1) {
+        callChecks.add(context);
+      }
+      return fn.apply(this, arguments);
+    };
+  }
+
+  verify() {
+    // Create array for errors
+    const errors = [];
+    for (const context of this.callChecks) {
+      // If functions have not been called exact times
+      if (context.actual !== context.exact) {
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+        errors.push(new AssertionError({
+          message,
+          actual: context.actual,
+          expected: context.exact,
+          operator: context.name,
+          startStackFn: context.stackTrace
+        }));
+      }
+    }
+    if (errors.length > 0)
+      throw errors;
+  }
+
+  report() {
+    try {
+      this.verify();
+    } catch (err) {
+      return err;
+    }
```

Comment:
I would invert report and verify and avoid the try-catch

Reply:
What do you mean invert?

Reply:
I mean have verify call report and throw, instead of having report calling verify with a try-catch.

Reply:
I'm a bit unsure of what you're suggesting. Do you mean that verify should call report, pass to it the errors and report should then throw?

Reply:
Yes, exactly.

Reply:
I originally intended `report()` to give the user a way to 'check up' on the functions without throwing an error. So now I have `report()` compiling an array of errors without calling `verify()`

Reply:
Let's make verify call report then

Reply:
I don't really understand, so do you reckon there should be a function that compiles an array of errors for the functions that have not been called the expected number of times, and two other functions that the array is passed to, one of them throwing the errors, and the other simply returning the array?

Reply:
Ah ignore that, I understand now